File: backend/src/config/controllers/sectionController.ts
import type { Request, Response } from 'express';
import * as sectionService from '../services/sectionService.js';

export const getAll = async (_req: Request, res: Response): Promise<void> => {
  try {
    const sections = await sectionService.findAll();
    res.json(sections);
  } catch (err) {
    console.error('getAll sections error:', err);
    res.status(500).json({ error: 'Internal Server Error' });
  }
};

export const getById = async (req: Request, res: Response): Promise<void> => {
  try {
    const section = await sectionService.findById(req.params.id);
    if (!section) { res.status(404).json({ error: 'Section not found' }); return; }
    res.json(section);
  } catch (err) {
    console.error('getById section error:', err);
    res.status(500).json({ error: 'Internal Server Error' });
  }
};

export const create = async (req: Request, res: Response): Promise<void> => {
  try {
    const { name, description } = req.body;
    if (!name) { res.status(400).json({ error: 'Field "name" is required' }); return; }
    const section = await sectionService.create({ name, description });
    res.status(201).json(section);
  } catch (err) {
    console.error('create section error:', err);
    res.status(500).json({ error: 'Internal Server Error' });
  }
};

export const update = async (req: Request, res: Response): Promise<void> => {
  try {
    const { name, description } = req.body;
    if (!name) { res.status(400).json({ error: 'Field "name" is required' }); return; }
    const section = await sectionService.update(req.params.id, { name, description });
    if (!section) { res.status(404).json({ error: 'Section not found' }); return; }
    res.json(section);
  } catch (err) {
    console.error('update section error:', err);
    res.status(500).json({ error: 'Internal Server Error' });
  }
};

export const remove = async (req: Request, res: Response): Promise<void> => {
  try {
    const deleted = await sectionService.remove(req.params.id);
    if (!deleted) { res.status(404).json({ error: 'Section not found' }); return; }
    res.json({ message: 'Section deleted successfully' });
  } catch (err) {
    console.error('delete section error:', err);
    res.status(500).json({ error: 'Internal Server Error' });
  }
};

File: backend/controllers/section_controller.go
package controllers

import (
	"net/http"

	"aura-erp/backend/models"
	"aura-erp/backend/services"

	"github.com/gin-gonic/gin"
)

func GetAllSections(c *gin.Context) {
	sections, err := services.GetAllSections()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal Server Error"})
		return
	}

	c.JSON(http.StatusOK, sections)
}

func GetSectionByID(c *gin.Context) {
	id := c.Param("id")

	section, err := services.GetSectionByID(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Section not found"})
		return
	}

	c.JSON(http.StatusOK, section)
}

func CreateSection(c *gin.Context) {
	var data models.SectionCreate
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if data.Name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Field \"name\" is required"})
		return
	}

	section, err := services.CreateSection(data)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal Server Error"})
		return
	}

	c.JSON(http.StatusCreated, section)
}

func UpdateSection(c *gin.Context) {
	id := c.Param("id")

	var data models.SectionUpdate
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if data.Name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Field \"name\" is required"})
		return
	}

	section, err := services.UpdateSection(id, data)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Section not found"})
		return
	}

	c.JSON(http.StatusOK, section)
}

func DeleteSection(c *gin.Context) {
	id := c.Param("id")

	err := services.DeleteSection(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Section not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Section deleted successfully"})
}
